### PR TITLE
Improve anchor tags

### PR DESCRIPTION
**Pull request by ian-r-rose**
Supersedes #1945.

This fixes a longstanding issue with relative links between documents that involve a header anchor tag. Does the following:
1. Our rendered markdown automatically generates header ids for markdown headers.
2. The link handler for the `rendermime-extension` now opens the relative document, and *then* calls a `node.querySelector` to search for an element with the corresponding id. If it finds that element, we scroll it into view.
3. Introduces a concept of `defaultRendered` for document widget factories, distinct from the default file type for that factory. This way we can have a file type (such as markdown), that opens with the text editor by default, but can also be opened with a `defaultRendered` factory. By default the `linkHandler` opens files with the `defaultRendered`.

### CODE REVIEW

**Comment by ian-r-rose** [comment]
I'd like to get this in for Beta 3, since it involves a minor version bump to `rendermime-interfaces`.

**Comment by jasongrout** [comment]
Taking a step back: Perhaps we should have the concept of default editor and default viewer? Should we insist that the default viewer be readonly? Right now I think we put the default editor as the first item on the `open with` list - should the default viewer be right under it? Right above it?

When double-clicking on a file, should we open it with the default editor or the default viewer? Should each filetype have a default default (i.e., should I use the default editor or the default viewer)?

**Comment by ian-r-rose** [comment]
I don't think we really have a concept of a read-only viewer right now, but we should (see the discussion in #3915  and #3251). We have a read-only model, but if the model is writable, then so is the viewer (leading to odd things like being able to "save" in the image viewer).

Right now the behavior of the `default` widget factory is that it always returns a widget factory. In this PR, I have made it so that the `defaultRendered` function returns the default rendered factory if it is available, and then falls back on the default factory if it is not. If we were to put the default editor and default viewer on equal footing we would want to rethink that logic. 

I agree about putting the default viewer right under the default editor in the `Open With` submenu.

**Comment by ian-r-rose** [comment]
This discussion about making a distinction between "viewers" and "editors" is significantly larger than the particular issue that this PR is trying to address. What do you think the best way forward on that is?

File: packages/docregistry/src/default.ts [hunk 1]
```diff
@@ -303,6 +303,9 @@ abstract class ABCWidgetFactory<T extends IDocumentWidget, U extends DocumentReg
     this._name = options.name;
     this._readOnly = options.readOnly === undefined ? false : options.readOnly;
     this._defaultFor = options.defaultFor ? options.defaultFor.slice() : [];
+    this._defaultRendered = options.defaultRendered ?
+                            options.defaultRendered.slice() :
+                            [];
```
Comment (afshin) [inline packages/docregistry/src/default.ts]: If you want to be cute:

```typescript
this._defaultRendered = (options.defaultRendered || []).slice();
```

File: packages/docregistry/src/registry.ts [hunk 2]
```diff
@@ -352,11 +363,41 @@ class DocumentRegistry implements IDisposable {
   }
 
   /**
-   * Get the default widget factory for an extension.
+   * Get the default rendered widget factory for a path.
+   *
+   * @param path - The path to for which to find a widget factory.
```
Comment (blink1073) [inline packages/docregistry/src/registry.ts]: the path for which?

**Comment by blink1073** [comment]
FWIW I like the name "rendered".   View implies read-only to me.  Also :+1: on listing this after the default entry in the Open with menu.

**Comment by ian-r-rose** [comment]
Rebased

**Comment by blink1073** [comment]
TSLint errors: https://travis-ci.org/jupyterlab/jupyterlab/jobs/402472768#L1206

**Comment by ian-r-rose** [comment]
Ah, right. Do you think it's worthwhile to run prettier in the `jlpm run integrity`?

**Comment by blink1073** [comment]
I think issue is that some things are not fixed by prettier.  Does prettier not run for you locally on git commits?

**Comment by ian-r-rose** [comment]
It does, but it was not picked up by my rebase.

I was more thinking: if the integrity test is checking prettier, then semantically I might expect that `jlpm run integrity` might apply it.

**Comment by blink1073** [comment]
Only if it is limited by git status: I wouldn't want to wait while it runs prettier on everything.

**Comment by ian-r-rose** [comment]
Fair enough.

**Comment by blink1073** [comment]
Were you planning to update the `Open With...` menu in this PR as well?

**Comment by ian-r-rose** [comment]
Yeah, I'll take a look at that today.

**Pull request merged**